chanfitness/chaneventstore_test.go

```diff
@@ -0,0 +1,421 @@
+						ShortChannelID: shortID,
+						IdentityPub:    privKey.PubKey(),
+					},
+				},
+				peernotifier.PeerOnlineEvent{PubKey: pubKey},
+			},
+			expectedEvents: []eventType{peerOnlineEvent},
+		},
+		{
+			name: "Channel opened, peer already online",
+			events: []interface{}{
+				channelnotifier.OpenChannelEvent{
+					Channel: &channeldb.OpenChannel{
+						ShortChannelID: shortID,
+						IdentityPub:    privKey.PubKey(),
+					},
+				},
+			},
+			onlinePeers:    [][33]byte{pubKey},
+			expectedEvents: []eventType{peerOnlineEvent},
+		},
+		{
+			name: "Channel known, peer offline",
+			events: []interface{}{
+				peernotifier.PeerOfflineEvent{PubKey: pubKey},
+			},
+			knownChannels:  []uint64{shortID.ToUint64()},
+			expectedEvents: []eventType{peerOfflineEvent},
+		},
+		{
+			name: "Channel known, peer offline, closed",
+			events: []interface{}{
+				peernotifier.PeerOfflineEvent{PubKey: pubKey},
+				channelnotifier.ClosedChannelEvent{
+					CloseSummary: &channeldb.ChannelCloseSummary{
+						ShortChanID: shortID,
+					},
+				},
+			},
+			knownChannels:  []uint64{shortID.ToUint64()},
+			expectedEvents: []eventType{peerOfflineEvent},
+		},
+		{
+			name: "Channel known, no events after close",
+			events: []interface{}{
+				channelnotifier.ClosedChannelEvent{
+					CloseSummary: &channeldb.ChannelCloseSummary{
+						ShortChanID: shortID,
+					},
+				},
+				peernotifier.PeerOfflineEvent{PubKey: pubKey},
+			},
+			knownChannels: []uint64{shortID.ToUint64()},
+		},
+		{
+			name: "Close unknown channel",
+			events: []interface{}{
+				channelnotifier.ClosedChannelEvent{
+					CloseSummary: &channeldb.ChannelCloseSummary{
+						ShortChanID: shortID,
+					},
+				},
+				peernotifier.PeerOfflineEvent{PubKey: pubKey},
+			},
+			skipCheck: true,
+		},
+		{
+			name: "Channel already known to store",
+			events: []interface{}{
+				channelnotifier.OpenChannelEvent{
+					Channel: &channeldb.OpenChannel{
+						ShortChannelID: shortID,
+						IdentityPub:    privKey.PubKey(),
+					},
+				},
+			},
+			knownChannels: []uint64{shortID.ToUint64()},
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+
+		t.Run(test.name, func(t *testing.T) {
+			// Create a store with the channels and online peers specified
+			// by the test.
+			store := &ChannelEventStore{
+				channels: make(map[uint64]*chanEventLog),
+				peers:    make(map[[33]byte]bool),
+				quit:     make(chan struct{}),
+			}
+			// Add known channels.
+			for _, c := range test.knownChannels {
+				store.addChannel(c, pubKey)
+			}
+
+			// Set known online peers.
+			for _, p := range test.onlinePeers {
+				store.peers[p] = true
+			}
+
+			// Create an events channel with sufficient buffer and add all the
+			// test's events.
+			events := make(chan interface{}, len(test.events))
+			for _, e := range test.events {
+				events <- e
+			}
+			close(events)
+
+			store.wg.Add(1)
+			store.monitorChannelEvents(events)
```

Comment:
This looks like it is using a lot if implementation knowledge of the store, making this test expensive to maintain. It needs to change if the internals of the store change. It is preferable to test it more as a black box.

Reply:
I think the solution to prevent peeking inside the event lists (implementation specific), would be to create a mock event log. But I am ok leaving it as is.